ledger/simulation/resources.go

```diff
@@ -0,0 +1,395 @@
+	if txn.Type != protocol.ApplicationCallTx {
+		return ResourceAssignment{}
+	}
+	return ResourceAssignment{
+		MaxAccounts:  proto.MaxAppTxnAccounts - len(txn.Accounts),
+		MaxAssets:    proto.MaxAppTxnForeignAssets - len(txn.ForeignAssets),
+		MaxApps:      proto.MaxAppTxnForeignApps - len(txn.ForeignApps),
+		MaxBoxes:     proto.MaxAppBoxReferences - len(txn.Boxes),
+		MaxTotalRefs: proto.MaxAppTotalTxnReferences - len(txn.Accounts) - len(txn.ForeignAssets) - len(txn.ForeignApps) - len(txn.Boxes),
+	}
+}
+
+func makeGlobalResourceAssignment(perTxnResources []ResourceAssignment) ResourceAssignment {
+	var globalResources ResourceAssignment
+	for i := range perTxnResources {
+		globalResources.MaxAccounts += perTxnResources[i].MaxAccounts
+		globalResources.MaxAssets += perTxnResources[i].MaxAssets
+		globalResources.MaxApps += perTxnResources[i].MaxApps
+		globalResources.MaxBoxes += perTxnResources[i].MaxBoxes
+		globalResources.MaxTotalRefs += perTxnResources[i].MaxTotalRefs
+	}
+	return globalResources
+}
+
+// IsEmpty returns true if the assignment has no resources.
+// NOTE: this is different from the traditional Go idiom of checking if the object is the zero value.
+func (a *ResourceAssignment) IsEmpty() bool {
+	return len(a.Accounts) == 0 && len(a.Assets) == 0 && len(a.Apps) == 0 && len(a.Boxes) == 0
+}
+
+func (a *ResourceAssignment) hasAccount(addr basics.Address, ep *logic.EvalParams, programVersion uint64) bool {
+	// nil map lookup is ok
+	_, ok := a.Accounts[addr]
+	if ok {
+		return true
+	}
+	if programVersion >= 7 { // appAddressAvailableVersion
+		for app := range a.Apps {
+			if ep.GetApplicationAddress(app) == addr {
+				return true
+			}
+		}
+	}
+	return false
+}
+
+func (a *ResourceAssignment) addAccount(addr basics.Address) bool {
+	if len(a.Accounts) >= a.MaxAccounts || len(a.Accounts)+len(a.Assets)+len(a.Apps)+len(a.Boxes) >= a.MaxTotalRefs {
+		return false
+	}
+	if a.Accounts == nil {
+		a.Accounts = make(map[basics.Address]struct{})
+	}
+	a.Accounts[addr] = struct{}{}
+	return true
+}
+
+func (a *ResourceAssignment) hasAsset(aid basics.AssetIndex) bool {
+	// nil map lookup is ok
+	_, ok := a.Assets[aid]
+	return ok
+}
+
+func (a *ResourceAssignment) addAsset(aid basics.AssetIndex) bool {
+	if len(a.Assets) >= a.MaxAssets || len(a.Accounts)+len(a.Assets)+len(a.Apps)+len(a.Boxes) >= a.MaxTotalRefs {
+		return false
+	}
+	if a.Assets == nil {
+		a.Assets = make(map[basics.AssetIndex]struct{})
+	}
+	a.Assets[aid] = struct{}{}
+	return true
+}
+
+func (a *ResourceAssignment) hasApp(aid basics.AppIndex) bool {
+	// nil map lookup is ok
+	_, ok := a.Apps[aid]
+	return ok
+}
+
+func (a *ResourceAssignment) addApp(aid basics.AppIndex, ep *logic.EvalParams, programVersion uint64) bool {
+	if len(a.Apps) >= a.MaxApps {
+		return false
+	}
+
+	if programVersion >= 7 { // appAddressAvailableVersion
+		appAddr := ep.GetApplicationAddress(aid)
+		// nil map lookup is ok
+		_, ok := a.Accounts[appAddr]
```

Comment:
Nice to have caught this case.  I suppose it's too hard to deal with the fact that you might have already failed the txn because too many Accounts were used, but some will turn out to be app accounts.

Reply:
maybe make the account limit be the accounts + app limits and then deduplicate apps from the accounts only at the end to make a final pass/fail

Reply:
I ended up taking @bbroder-algo's idea in 3fa219d44e90b309f33c7f5ee9756064818ddf57